src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs

```diff
@@ -581,6 +561,31 @@ private CSharpCodeWriter RenderPreWriteStart()
             return RenderPreWriteStart(Writer, Context);
         }
 
+        private void RenderChunkedWriteLiteral(string text)
+        {
+            var charactersRendered = 0;
+
+            // Render the string in pieces to avoid Roslyn OOM exceptions at compile time:
+            // https://github.com/dotnet/roslyn/issues/7398
+            while (charactersRendered < text.Length)
+            {
+                if (Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput)
+                {
+                    RenderPreWriteStart();
+                }
+
+                var charactersToSubstring = Math.Min(MaxStringLiteralLength, text.Length - charactersRendered);
+                var textToRender = text.Substring(charactersRendered, charactersToSubstring);
+                Writer.WriteStringLiteral(textToRender);
+                charactersRendered += textToRender.Length;
+
+                if (Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput)
+                {
+                    Writer.WriteEndMethodInvocation();
+                }
+            }
+        }
+
         public static CSharpCodeWriter RenderPreWriteStart(CSharpCodeWriter writer, CodeGeneratorContext context)
```

Comment:
Remove this `static` method and inline it into the unhelpful `RenderPreWriteStart()` method. Better if that's renamed to something meaningful e.g. `RenderStartWriteLiteralMethodInvocation()`. "pre write" is meaning-challenged.